### ISSUE TITLE
Have a base "zero ITensor" for summing ITensors?

### ISSUE THREAD

**Issue by emstoudenmire** [ITensor/ITensors.jl#340]
A pattern that can come up often is that one wants to sum up a set of ITensors in a for loop, but the code inside the loop is rather complicated and the tensor accumulating the results doesn't exist yet. For example (it's a fake example, so no need to press into the details):

```
Q,R = qr(v[1],s[1])
T = alpha*Q*w[1]

for n=2:N
  Q, R = qr(v[n],s[n])
  T += alpha*Q*w[n]
end
```

As can be seen above, the complicated code logic inside the loop has to be repeated in front of the loop to initialize T, causing code duplication and possible bugs etc.

In the C++ version of ITensor we allow T to be initialized as `auto T = ITensor()` and then other ITensors are allowed to be added into T, so that it acts as a kind of "universal zero" tensor. But maybe in Julia we can do something better to solve this problem? E.g. we could allow T to be initialized as `nothing` and then add to it?

**Comment by emstoudenmire** [comment]
I tried it just now, and it requires very little code to implement:

```
Base.:+(A::Nothing,B::ITensor) = copy(B)
Base.:+(A::ITensor,B::Nothing) = copy(A)
```

Of course the question is whether it's a good idea.

**Comment by mtfishman** [comment]
Yes, I was trying to think of a good interface for this. I think I prefer an interface where it is explicit that you are creating an ITensor with a special storage type. I would prefer to preserve plain constructors like `ITensor()` for simpler meanings, for example right now it creates a Dense ITensor with no indices and one element of zero, which is expected in many ways from other similar Dense constructors.

I don't really like using `nothing` for that purpose. There may be functions which return `nothing` or an ITensor, where you want operations with that `nothing` to fail and in order to catch that an ITensor was not returned.

It seems like we would want a special `Zero` storage type like in C++, the purpose of which is that it acts as an additive identity (and I guess could also act as multiplicative zero as well). We could give it an explicit sparse constructor function like `zeroITensor`, which matches with functions like `diagITensor` to create an ITensor with a `Diag` storage.

**Comment by emstoudenmire** [comment]
Good point about the `nothing` approach leading to unwelcome surprises. I agree we should exercise caution here.

One thing that could be confusing, though, is if we have two kinds of zero ITensors both of which have no indices, but only one which is allowed to add to other ITensors with indices. Or would it be allowed to add both an ITensor() and a zeroITensor() to another ITensor?

**Comment by mtfishman** [comment]
The way I was picturing it (and how it is designed now), `ITensor() = ITensor(0)`, so it is really just a scalar ITensor with no indices that happens to have a single value of zero. In that case, it wouldn't make sense to do something like `ITensor() + ITensor(i, j)`. That would be what distinguishes it from an ITensor constructed with `zeroITensor()`, which can be added to an ITensor with any Index structure. To me it makes sense to distinguish those cases, since adding an ITensor with no indices to one with indices is very unique behavior that I think someone should have to ask for explicitly.

**Comment by emstoudenmire** [comment]
Ok great, that makes sense that the distinction is one has to ask for a zeroITensor(), i.e. actually type that out to declare intent.

**Comment by mtfishman** [comment]
Another question is if `zeroITensor` should be used in the following case:
```julia
i = Index(2)
A = zeroITensor(i, i')
A[i=>1, i'=>1] = 1.0 + 2.0im
```
i.e. the case where you don't want the storage initialized until you start setting elements. I find it a bit strange, because in the use case `zeroITensor() + randomITensor(i, i')` it acts like it has any number of indices, while in the element-setting case it needs to be defined with indices, but maybe that is ok.

**Comment by mtfishman** [comment]
I also wanted to note that in theory we could do the following:
```julia
A = zeroITensor()
i = Index(2)
A[i=>1, i'=>1] = 1.0 + 2.0im
```
where the call to `setindex!` determines that the new ITensor will have indices `(i, i')`. I noticed that this might be useful for more generic code in places like `op`, where it sometimes creates ITensors with pairs of site indices `(s', dag(s))` and sometimes makes ITensors with just one site index `s`.

**Comment by emstoudenmire** [comment]
Closed by #348
